connections/src/main/java/com/stripe/android/connections/di/ComponentHolder.kt

```diff
@@ -0,0 +1,12 @@
+package com.stripe.android.connections.di
+
+/**
+ * Global instance that holds application component with entire DI graph.
+ */
+object ComponentHolder {
+  val components = mutableSetOf<Any>()
+
+  inline fun <reified T> component(): T = components
+    .filterIsInstance<T>()
+    .single()
+}
```

Comment:
Took this idea from the Anvil sample - https://github.com/square/anvil/blob/main/sample/scopes/src/main/java/com/squareup/scopes/ComponentHolder.kt

Reply:
It seems like we will always be adding to `components` without ever deleting it, I feel like this might be intrusive for the client app - if they say need to use `ConnectionsSheet` multiple times by restarting the same Activity over and over, even though the `ConnectionSheet` instance is GC-ed, we ended up saving the unused graphs created here. (That's why we used the [weakmap](https://github.com/stripe/stripe-android/blob/e132f0eb6f84b037ff8ab35234bb0af14c8374ec/stripe-core/src/main/java/com/stripe/android/core/injection/WeakMapInjectorRegistry.kt#L24) solution to bind the graph to the API object itself in Payment, when `PaymentLauncher` instance is GC-ed, its corresponding dagger graph is also gone)

Reply:
The `Component` & `Subcomponent` solution currently used in Payment makes sure that for the same API object, there is some shared instances saved in `Component` (e.g network stack with a publishable key), when the API object starts activities, the subcomponent provides activity specific dependencies.

It seems that the approach in this PR, ensures that the `Component` is created on `Application#onCreate` and shared across all API objects created later, I think the benefit of having this approach is 
* We no longer need to worry about activity being killed, as everything is saved in Application
* It might be faster to instantiate API instances, as part of them is already there during `Application#onCreate`


However I think there are also two concerns on this approach


* It is possible that the client will never execute our code, but we still occupies the memory regardless, this might become more intrusive given that the size `Component` might grow over time
* Saving a shared `Component` across multiple API instances seems a bit dangerous, we'll either need to make sure the instances saved inside it are stateless or let the newly created API instance alter the state of the global component(e.g in Identity we don't use publisableKey, but always use a temporary [ephemeralKey](https://github.com/stripe/stripe-android/blob/dcc48ef861fc596009390fb3c88812e1464d84a3/identity/src/main/java/com/stripe/android/identity/IdentityVerificationSheet.kt#L56) in the present method, that key needs to be propagated to the network stack, which might be saved in `Component`). I think this can be done by `@Named` or `@Provider` like the [publishablekey](https://github.com/stripe/stripe-android/blob/efbcd515416099713afe8c6ac7307cd97497e0a7/payments-core/src/main/java/com/stripe/android/networking/StripeApiRepository.kt#L128).

Given the pros and cons, not sure if the benefit outweighs the potential maintainance cost... wdyt?

Reply:
> It seems like we will always be adding to components without ever deleting it, would this work if the need to use ConnectionsSheet multiple times therefore need different dagger graph content?

- This should only host the AppComponent / Broader scope Component. `ConnectionsSheetComponent` would be a subcomponent of it. On process recreation, that parent component (that was destroyed) will get recreated. 

> It is possible that the client will never execute our code, but we still occupies the memory regardless, this might become more intrusive given that the size Component might grow over time

- Agreed. The idea would be to keep the minimum necessary App-scoped dependencies, and use subcomponents to handle narrower scopes. 

> Saving a shared Component across multiple API instances seems a bit dangerous, we'll either need to make sure the instances saved inside it are stateless or let the newly created API instance alter the state of the global component(e.g in Identity we don't use publisableKey, but always use a temporary [ephemeralKey]

- This would bring the ability to share dependencies at a scope level, but it'd be up to the developer to decide the scope of a given dependency. In your mentioned example,  a network client could be app-scoped, and we could also have a stateful app-scoped ephemeralKey provider. Or have both `IdentitySheet` scoped as we do right now, but being `IdentitySheetComponent` a subcomponent of an almost empty `AppComponent` that would be just used to handle graph recreation.